### ISSUE TITLE
Unobserved task exception when server is not reachable

### ISSUE THREAD

**Issue by mm-fee** [grpc/grpc#24421]
### What version of gRPC and what language are you using?
Grpc 2.32.0, Google.Protobuf 3.13.0, C# 7.3

### What operating system (Linux, Windows,...) and version?
Windows 10 x64 2004

### What runtime / compiler are you using (e.g. python version or version of gcc)
.NET Core SDK 3.1.403

### What did you do?
I created a new program just consisting of this one class and the proto file below.
```c#
using System;
using System.Text;
using System.Threading.Tasks;

using Google.Protobuf;

using Grpc.Core;

namespace TaskExceptionDemo
{
    internal class Program
    {
        public static async Task Main()
        {
            TaskScheduler.UnobservedTaskException += (sender, e) => Console.WriteLine($"Unobserved Exception: {e.Exception}");

            try
            {
                await Task.Run(async () =>
                                   {
                                       var channel = new Channel("127.0.0.1", 20000, ChannelCredentials.Insecure);
                                       var client = new TaskExceptionDemoService.TaskExceptionDemoServiceClient(channel);

                                       using var call = client.DoSomething();

                                       var requestModel = new RequestModel
                                                              {
                                                                  RequestData = ByteString.CopyFrom("Hello World", Encoding.UTF8)
                                                              };

                                       try
                                       {
                                           await call.RequestStream.WriteAsync(requestModel);
                                           await call.RequestStream.CompleteAsync();
                                       }
                                       catch (RpcException e)
                                       {
                                           if (e.StatusCode == StatusCode.Unavailable)
                                           {
                                               Console.WriteLine("Server not responding");
                                           }
                                           else
                                           {
                                               Console.WriteLine($"Unknown Error: {e.Message}");
                                           }
                                       }
                                       finally
                                       {
                                           await channel.ShutdownAsync();
                                       }
                                   });

                GC.Collect();
                Console.WriteLine("Press Enter to finish.");
                Console.ReadLine();
            }
            catch (Exception e)
            {
                Console.WriteLine($"Unexpected exceptions from tasks. {e}");
            }
        }
    }
}
```

```proto
syntax = "proto3";

option csharp_namespace = "TaskExceptionDemo";

service TaskExceptionDemoService {
	
	rpc DoSomething (stream RequestModel) returns (stream ReplyModel);
}

message RequestModel {
	bytes requestData = 1;
}

message ReplyModel {	
	bytes replyData = 1;

}
```

### What did you expect to see?
The following console output:
Server not responding
Press Enter to finish.

### What did you see instead?

```c#
Server not responding
Press Enter to finish.
Unobserved Exception: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Status(StatusCode="Unavailable", Detail="failed to connect to all addresses", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1602749024.126000000","description":"Failed to pick subchannel","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\client_channel.cc","file_line":4134,"referenced_errors":[{"created":"@1602749024.126000000","description":"failed to connect to all addresses","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\lb_policy\pick_first\pick_first.cc","file_line":398,"grpc_status":14}]}"))
 ---> Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="failed to connect to all addresses", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1602749024.126000000","description":"Failed to pick subchannel","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\client_channel.cc","file_line":4134,"referenced_errors":[{"created":"@1602749024.126000000","description":"failed to connect to all addresses","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\lb_policy\pick_first\pick_first.cc","file_line":398,"grpc_status":14}]}")
   --- End of inner exception stack trace ---

```

### Anything else we should know about your project / environment?
It seems not to happen when the client code is not running in a task.
I think the problem maybe related to #17458 but was obviously not fixed by #21202

**Comment by jskeet** [comment]
I've reproduced this with a simpler piece of code:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Google.Cloud.Firestore.V1" Version="2.1.0" />
  </ItemGroup>
</Project>
```

Program.cs:

```csharp
using Google.Cloud.Firestore.V1;
using Grpc.Core;
using System;
using System.Threading.Tasks;

class Program
{
    static async Task Main()
    {
        TaskScheduler.UnobservedTaskException += (sender, args) =>
            Console.WriteLine($"Unhandled exception: {args.Exception.InnerException}");

        // Deliberately invalid endpoint
        ChannelBase channel = new Channel("10.1.1.1", ChannelCredentials.Insecure);
        var client = new Firestore.FirestoreClient(channel);

        // Create a streaming request, which will of course fail.
        var stream = client.Listen();
        Task task = stream.RequestStream.WriteAsync(new ListenRequest());

        try
        {
            // Observe the failure normally
            await task;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Caught {e} normally");
        }

        // Provoke the garbage collector to find the unobserved exception.
        Console.WriteLine();
        Console.WriteLine("Waiting for finalizers");

        // Wait for any failed tasks to be garbage collected
        //await Task.Delay(5000);
        GC.Collect();
        GC.WaitForPendingFinalizers();
    }
}
```

Run with `dotnet run -c Release`. Output:

```text
Caught Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="failed to connect to all addresses", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1603103650.880000000","description":"Failed to pick subchannel","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\client_channel.cc","file_line":4090,"referenced_errors":[{"created":"@1603103650.880000000","description":"failed to connect to all addresses","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\lb_policy\pick_first\pick_first.cc","file_line":394,"grpc_status":14}]}")
   at Program.Main() in C:\users\jonskeet\Issues\google-cloud-dotnet\Issue5462\ConsoleListener\Program.cs:line 24 normally

Waiting for finalizers
Unhandled exception: Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="failed to connect to all addresses", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1603103650.880000000","description":"Failed to pick subchannel","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\client_channel.cc","file_line":4090,"referenced_errors":[{"created":"@1603103650.880000000","description":"failed to connect to all addresses","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\ext\filters\client_channel\lb_policy\pick_first\pick_first.cc","file_line":394,"grpc_status":14}]}")
```

**Comment by anon17** [comment]
I had to run it several times before it fails
```cs
[TestMethod]
public async Task TestUnobservedTaskException()
{
	TaskScheduler.UnobservedTaskException += TaskScheduler_UnobservedTaskException;
	var chan = new Channel("localhost:100", ChannelCredentials.Insecure);
	var client = new TaskExceptionDemoService.TaskExceptionDemoServiceClient(chan);
	for (int i = 0; i < 3; i++)
		try
		{
			var call = client.DoSomething();
			using (call)
			{
				var req = call.RequestStream;
				await req.WriteAsync(new RequestModel());
				Assert.Fail("huh?");
			}
		}
		catch (RpcException err)
		{
			Debug.WriteLine(err.Status.StatusCode);
		}
	await chan.ShutdownAsync();
	GC.Collect();
	GC.WaitForPendingFinalizers();
	TaskScheduler.UnobservedTaskException -= TaskScheduler_UnobservedTaskException;
}

static void TaskScheduler_UnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs e)
{
	Debug.WriteLine("unobserved exception");
}
```

**Comment by anon17** [comment]
To make the test run faster connect to 0.0.0.0

**Comment by jtattermusch** [comment]
This was fixed by https://github.com/grpc/grpc/pull/29582. The fix will be in the 2.46.2+ patch release (which will be released soon).
